src/carnot/core/data/auto_retrieval/_internal/index_management.py

```diff
@@ -0,0 +1,1037 @@
+                texts_batch.append(text)
+                metas_batch.append(metadata)
+
+            if len(ids_batch) >= self._batch_size:
+                flush()
+
+        flush()
+
+    def get_documents(self, doc_ids: Sequence[str]) -> Sequence[Mapping[str, Any]]:
+        """Retrieve documents (text + metadata) by ID from Chroma."""
+        if not doc_ids:
+            return []
+
+        results = self._collection.get(
+            ids=list(doc_ids),
+            include=["documents", "metadatas"],
+        )
+
+        found_ids = results.get("ids", [])
+        found_texts = results.get("documents", [])
+        found_metas = results.get("metadatas", [])
+
+        out: List[Mapping[str, Any]] = []
+        for i, doc_id in enumerate(found_ids):
+            out.append(
+                {
+                    "id": doc_id,
+                    "text": found_texts[i],
+                    "metadata": found_metas[i] if found_metas else {},
+                }
+            )
+        return out
+
+    def list_document_ids(self) -> Sequence[str]:
+        """List all stored chunk IDs."""
+        results = self._collection.get(include=[])
+        return results.get("ids", [])
+
+    # ---------- BaseVectorIndex API ----------
+
+    def query(self, query_embedding: Sequence[float], top_k: int, include: Sequence[str] = ["documents", "metadatas", "distances"], filters: Optional[Mapping[str, Any]] = None) -> Tuple[Sequence[str], Sequence[str], Sequence[Mapping[str, Any]], Sequence[float]]:
+        """
+        Query Chroma for nearest neighbors given a precomputed embedding.
+
+        Returns a list of (id, distance) pairs.
+        """
+        logger.info(f"ChromaVectorIndex: querying with top_k={top_k}, filters={filters}")
+        results = self._collection.query(
+            query_embeddings=[list(query_embedding)],
+            n_results=top_k,
+            include=include,
+            where=filters,
+        )
+
+        ids = results.get("ids", [[]])[0]
+        documents = results.get("documents", [[]])[0]
+        metadatas = results.get("metadatas", [[]])[0]
+        distances = results.get("distances", [[]])[0]
+
+        if not ids:
+            return ([], [], [], [])
+
+        return (ids, documents, metadatas, distances)
+
+
+# ---- Simple keyword index stub (kept minimal for now) ----
+
+class ChromaKeywordIndex(BaseKeywordIndex):
+    """Placeholder keyword / inverted index."""
+
+    def __init__(self, config: Config) -> None:
+        pass
+
+    def add_documents(self, docs: Iterable[Mapping[str, Any]]) -> None:
+        """Index raw text and tokens for lexical retrieval (not implemented yet)."""
+        return
+
+    def query(self, terms: Sequence[str], top_k: int) -> Sequence[Tuple[str, float]]:
+        """Return documents matching the given keywords (empty for now)."""
+        return []
+
+
+# ---- Structured metadata table with uniform schema ----
+
+class TableMetadataStore(BaseMetadataStore):
```

Comment:
I don't think we want to roll our own in-memory metadata store. Realistically, we have two options:
- Use Chroma's metadata store and integrated querying (easy, but not extensible)
- Use something like DuckDB + Parquet that allows for optimizations. We can then manually pass an "allow-list" into Chroma if pre-filtering

We will eventually need 2). We can start with 1), but the in-memory table is definitely not sufficient.